File: easybluetoothble/src/main/java/com/srise/easybluetoothble/impl/BleSendDataEngine.java
package com.srise.easybluetoothble.impl;

import android.util.Log;

import com.srise.easybluetoothble.Interface.ISendDataEngine;
import com.srise.easybluetoothble.UbtBlueToothConnManager;
import com.srise.easybluetoothble.bean.BaseMessage;

import java.util.ArrayList;
import java.util.List;


public class BleSendDataEngine implements ISendDataEngine {
    private static final String TAG = "BleSendDataEngine";

    /**
     * 如果数据帧长度超过20则分包
     *
     * @param data
     * @return
     */
    @Override
    public List<byte[]> splitDataPacket(byte[] data) {
        Log.d(TAG, " splitDataPacket");

        List<byte[]> bytesList = new ArrayList<>();
        int splitSize = (int) (Math.ceil(data.length / 20.f));
        int remainder = data.length % 20;

        for (int index = 0; index < splitSize; index++) {
            byte[] partData;

            if (index == splitSize - 1) {
                partData = new byte[remainder];
                System.arraycopy(data, (index * 20), partData, 0, remainder);

            } else {
                partData = new byte[20];
                System.arraycopy(data, (index * 20), partData, 0, 20);
            }

            bytesList.add(partData);
        }

        return bytesList;
    }

    /**
     * 发送数据帧
     *
     * @param bleSendMessage
     */
    @Override
    public void sendData(BaseMessage bleSendMessage) {
        Log.d(TAG, " sendData");

        byte[] frameData = bleSendMessage.getMessageBody();
        List<UbtBlueToothConnManager.DataToNotifyPhoneEntity> listBytes = new ArrayList<>();

        if (frameData.length > 20) {
            List<byte[]> bytesArray = splitDataPacket(frameData);

            for (byte[] bytes : bytesArray) {
                listBytes.add(createEntity(bleSendMessage, bytes));
            }
        } else {
            listBytes.add(createEntity(bleSendMessage, frameData));
        }

        if (bleSendMessage.getFinishCallBack() != null) {
            bleSendMessage.getFinishCallBack().readyToSend(listBytes);
        }
    }

    private UbtBlueToothConnManager.DataToNotifyPhoneEntity createEntity(BaseMessage bleSendMessage, byte[] bytes) {
        UbtBlueToothConnManager.DataToNotifyPhoneEntity entity = new UbtBlueToothConnManager.DataToNotifyPhoneEntity();
        entity.setData(bytes);
        entity.setCharacteristic(bleSendMessage.getCharacteristic());
        entity.setDevice(bleSendMessage.getDevice());
        return entity;
    }
}

File: ktlionble/src/main/java/c/example/dell/ktlionble/impl/BleSendDataEngine.kt
package c.example.dell.ktlionble.impl

import android.util.Log
import c.example.dell.ktlionble.bean.BaseMessage
import c.example.dell.ktlionble.bean.DataToNotifyPhoneEntity
import c.example.dell.ktlionble.inter.ISendDataEngine
import java.util.*
import kotlin.math.ceil

class BleSendDataEngine : ISendDataEngine {
    companion object {
        const val TAG = "BleSendDataEngine"
    }


    override fun splitDataPacket(data: ByteArray?): List<ByteArray?>? {
        Log.d(TAG, " splitDataPacket")

        val bytesList: MutableList<ByteArray> = ArrayList()
        val splitSize = ceil(data!!.size / 20f.toDouble()).toInt()
        val remainder = data.size % 20

        for (index in 0 until splitSize) {
            var partData: ByteArray

            if (index == splitSize - 1) {
                partData = ByteArray(remainder)
                System.arraycopy(data, index * 20, partData, 0, remainder)
            } else {
                partData = ByteArray(20)
                System.arraycopy(data, index * 20, partData, 0, 20)
            }

            bytesList.add(partData)
        }

        return bytesList
    }

    override fun sendData(bleSendMessage: BaseMessage) {
        Log.d(TAG, " sendData")

        val frameData: ByteArray = bleSendMessage.mMessageBody!!
        val listBytes: MutableList<DataToNotifyPhoneEntity?> = ArrayList<DataToNotifyPhoneEntity?>()

        if (frameData.size > 20) {
            val bytesArray = splitDataPacket(frameData)

            for (bytes in bytesArray!!) {
                listBytes.add(createEntity(bleSendMessage, bytes!!))
            }
        } else {
            listBytes.add(createEntity(bleSendMessage, frameData))
        }

        bleSendMessage.mFinishCallBack!!.readyToSend(listBytes)
    }

    private fun createEntity(bleSendMessage: BaseMessage, bytes: ByteArray): DataToNotifyPhoneEntity? {
        val entity: DataToNotifyPhoneEntity = DataToNotifyPhoneEntity()
        entity.mData = bytes
        entity.mCharacteristic = bleSendMessage.mCharacteristic!!
        entity.mDevice = bleSendMessage.mDevice!!

        return entity
    }
}